### ISSUE TITLE
How to extract cam during training without creating issues to the bakpropagation

### ISSUE THREAD

**Issue by PietroManganelliConforti** [jacobgil/pytorch-grad-cam#488]
Hi,

I'm implementing the following custom function:

> def return_cam_from_model(model, target_layer, batch, targets , cam_name = "gradcam"):
> 
>     model.eval() # da rimettere in .train() se nel training loop
> 
>     cam = None
> 
>     if cam_name == "gradcam":
>         cam = GradCAM(model=model, target_layers=target_layer, use_cuda=False)
> 
>     elif cam_name == "scorecam":
>         cam = ScoreCAM(model=model, target_layers=target_layer, use_cuda=True) # CONTROLLARE CHE NON VA
> 
>     elif cam_name == "fullcam":
>         cam = FullGrad(model=model, target_layers=target_layer, use_cuda=True)
> 
>     elif cam_name == 'gradcamplspls':
>         cam = GradCAMPlusPlus(model=model, target_layers=target_layer, use_cuda=True)
>     
>     elif cam_name == 'xgradcam':
>         cam = XGradCAM(model=model, target_layers=target_layer, use_cuda=True)
>         
>     elif cam_name == 'eigencam':
>         cam = EigenCAM(model=model, target_layers=target_layer, use_cuda=True)
>      
>     elif cam_name == 'eigengradcam':
>         cam = EigenGradCAM(model=model, target_layers=target_layer, use_cuda=True)
>         
>     elif cam_name == 'layercam':
>         cam = LayerCAM(model=model, target_layers=target_layer, use_cuda=True)
> 
>     elif cam_name == 'fullgrad':
>         cam = FullGrad(model=model, target_layers=target_layer, use_cuda=True)
>      
>     elif cam_name == 'hirescam':
>         cam = HiResCAM(model=model, target_layers=target_layer, use_cuda=True)
>         
>     elif cam_name == 'gradcamelementwise':
>         cam = GradCAMElementWise(model=model, target_layers=target_layer, use_cuda=True)
>     
>     elif cam_name == 'NOXAI':
>         return 0
> 
>     else:
>         raise Exception("Cam name", cam_name, "not recognized")
>     
>     targets = [ClassifierOutputTarget(i) for i in targets]
> 
> 
>     ret_cam = cam(input_tensor=batch, targets=targets)
> 
>     model.train()
> 
>     return ret_cam
>     
    
we have seen that model.train() before the CAM execution produces a wrong cam explanation, so we put this function after in the code. Therefore, because the cam execution probably is affected by the model training, do you know if inside the cam execution there is something that may affect the model training/backward?

**Comment by PietroManganelliConforti** [comment]
I'm trying to get tensors with gradient after the library execution

**Comment by m-aryayi** [comment]
> I'm trying to get tensors with gradient after the library execution

Hi,
What exactly do you mean by that?

**Comment by jinCode-gao** [comment]
> I'm trying to get tensors with gradient after the library execution

I also want to do the same thing: obtain the cam matrix in the tensor version. Have you got the result?
